### DISCUSSION TITLE
Cannot figure out batteryMask

### DISCUSSION THREAD

**Discussion by andyfsimon** [mKeRix/room-assistant#588] [Q&A] [answered]
Hello,
I am trying to get the battery level from both my phones (Android with HA companion app sending iBeacons) and an actual BLE tag (I think RDL51822). I`ve tried to understand the logic of the docs about it, but I get very random numbers back.

What does exactly mean "The mask operates on a 32bit value with major as the high two bytes and minor as the low two bytes." ?
I have the Major & Minor numbers for each tag, how do I build the mask for it?

**Comment by PeteBa** [comment, marked as answer]
First question, do you know if your devices transmits the battery level in the iBeacon advertisement message ? Only certain devices do this and even then you generally need to tell the device that is what you want as it is a non-standard mode. The device documentation should be able to tell you if it is possible and how to get the iBeacon into that mode. I kniow the guys are working on room-assistant integration with companion apps but didnt think they provided battery information in the advertisement message. But I could be out of date on that.

I have five Jaylee ibeacons and they all support sending battery level in the advertisement. Their Jaylee docs explain how to "turn on the feature" and how the device will send the battery level. In my case, the devices put the battery level into the **first (or high) byte of the two-byte Minor field**. Having been told that, I can then configure two aspects of room-assistant so it knows how to interpret the iBeacon's address and pull out the battery info.

The first step is to tell room-assistant to ignore - mask out - the high byte of the Minor field from my device's addresses. Otherwise the changing battery level would cause room-assistant to think it was seeing a new device. You configure the address mask with the `minorMask` and `majorMask` options in local.yml. For me, I have the `majorMask` set to "0xFFFF" and the `minorMask` set to "0x00FF". If the docs said the battery level was on the low or second byte of the Major field then I would set the `majorMask` as 0xFF00 and `minorMask` as 0xFFFF. Basically, in the world of hexidecimal numbers, the 0's tell room-assistant which parts of the address data to ignore as it is being used for other purposes.

The second bit is to tell room-assistant where the battery information is found using the `batteryMask` option. Now this may seem like a bit of deja-vu but you cant guarantee that the byte you masked out from being used in the address is the same that holds the battery. Some devices squeeze more than just a battery level into a byte (i.e. they use half a byte to send a 0-15 level and the other half byte to send something else) whereas others use the full byte to send a 0-100 or even 0-255 level. So you need to also tell room-assistant exactly where the battery information is. In my case, it is, in fact, the full, high byte of the minor field. So the `batteryMask` I use is 0x0000FF00. The first two bytes on the left are the Major Field and the last two bytes on the right are the Minor Mask bytes. If the device happened to be transmitting the battery level in the bottom half-byte of the major field then the `batteryMask` would be "0x000F0000. Again the F's go where you want to keep the information and the 0's where you want it ignored. 

I hope that helps a bit and let me know if needs some clarification.

**Reply by andyfsimon** [reply]
Hi PeteBa, that is a very enlightening and precise answer, thank you!
Ok, so I tried to go through the "documentation" I could find for the RDL51822, which turns out to be likely an Nordic nRF51822 of sort. I bought it cheap on Aliexpress so can`t ask too much...
I did install an app from Nordic that does show some of the characteristics of the beacons around me but when I look at the list of flags and services I don`t see anything related to battery level. I suppose then that those beacons don`t rely that information.

Thank you very much for your answer!

**Reply by frankiej911** [reply]
Hello, I have the same iBeacon, it's the nRF51822 from Radioland (RDL). You can find the manual here: https://fccid.io/2AGUT-NRF52810/User-Manual/15-NRF52810B3-UserMan-US-4587802.pdf. With the app RLBeacon you can find the battery level for your beacons and you can also configure some RSSI/Major/Minor fields, and that's working for me. There is a topic on page 12/16 telling where you can find the battery level in the broadcast massage:

![image](https://user-images.githubusercontent.com/45697454/112903121-674ee180-90e7-11eb-980d-bd76198de374.png)

Now we must convert it to the right batteryMask for Room Assistant....

**Comment by PeteBa** [comment]
@frankiej911 , hmm, had a quick look at the linked doc and not quite sure that I fully understand. It seems to suggest that the beacon can be configured to send battery power (interpreted as level) in the field usually associated with the RRSI value. This would be odd as the RSSI field is part of the iBeacon spec and fairly important in determining distance. Have you confirmed that Is correct and not just a translation error? Nonetheless, `batteryMask` operates on the Major and Minor fields only and not byte 29.

**Reply by frankiej911** [reply]
Hello PeteBa, I had read that section too, but I don’t know and didn’t understand the document either about the battery and how to convert the major/minor/RSSI fields to the correct values to enter. 

I’ve ignored the RSSI/Major/Minor converting and enter the normal values of Major 1 Minor 1 and RSSI 1m to be -59.

This app, but also Room Assistant and some other general bluetooth apps shows the correct UUID-major-minor and the signal strength. The distance in Room Assistant doesn’t changed very quickly, so I don’t understand the section about the RSSI/BattPower combination...

https://user-images.githubusercontent.com/45697454/112952654-0651f880-913d-11eb-946d-828f4cab7620.png
https://user-images.githubusercontent.com/45697454/112952670-094ce900-913d-11eb-8599-d576ee91a7bb.png

Eidt: pictures in link, they were too big in this message.

**Comment by joaopdmdias** [comment]
Hi @PeteBa ,

I have the same issue, I would like to know how can I track battery levels from the RDL51822. Did you manage to solve it?
